### PR TITLE
Warn about options that are invalid for a given workbox-build mode

### LINKED ISSUE #708: Warn when injectManifest({runtimeCaching, navigateFallback}) is called
**Library Affected**:
*workbox-build*

There are a few options that could be passed in to `workbox-build.injectManifest()` (either directly, or via `workbox-cli` or `workbox-webpack-plugin`) that don't make any sense, since in that mode, we're only replacing the contents of `precache([])`, and not making any other changes to the `swSrc` file. The ones that I can think of include:

- `runtimeCaching`
- `navigateFallback`
- `navigateFallbackWhitelist`

but there might be others—as part of #695 I might come across more.

Instead of silently ignoring them, leading to developer confusion when they don't have any effect, we should throw a fatal exception and explain why they shouldn't be used in the `injectManifest()` mode.

### PR DESCRIPTION
**Pull request by jeffposnick**
R: @addyosmani @gauntface

Fixes #708

The decision reached in that issue was to log a warning, but not fail the build.

### CODE REVIEW

File: packages/workbox-build/src/lib/utils/warn-about-config.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * A helper to find and log any configuration options that were set which are
+ * not valid for a given usage mode.
+ *
+ * @param {Array<String>} blacklist The names of invalid configuration options.
+ * @param {Object<String,String>} fullConfig The configuration options to check.
+ */
+module.exports = (blacklist, fullConfig) => {
+  const invalidConfig = Object.keys(fullConfig)
+    .filter((configKey) => blacklist.indexOf(configKey) !== -1);
```
Comment (gauntface) [inline packages/workbox-build/src/lib/utils/warn-about-config.js]: Could you use includes?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

File: packages/workbox-build/src/lib/utils/warn-about-config.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * A helper to find and log any configuration options that were set which are
+ * not valid for a given usage mode.
+ *
+ * @param {Array<String>} blacklist The names of invalid configuration options.
+ * @param {Object<String,String>} fullConfig The configuration options to check.
+ */
+module.exports = (blacklist, fullConfig) => {
+  const invalidConfig = Object.keys(fullConfig)
```
Comment (gauntface) [inline packages/workbox-build/src/lib/utils/warn-about-config.js]: Nit: Rename to invalidConfigKeys (Just reads better for if (invalidConfigKeys.length > 0)

File: packages/workbox-build/src/lib/utils/warn-about-config.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    .filter((configKey) => blacklist.indexOf(configKey) !== -1);
+
+  if (invalidConfig.length > 0) {
+    logHelper.warn(`These Workbox configuration options are not valid for ` +
+      `the current mode, and have been ignored: ${invalidConfig}`);
```
Comment (gauntface) [inline packages/workbox-build/src/lib/utils/warn-about-config.js]: Nit: Swap "have been" with "will be" as I'm assuming this log will be at the start of the build task (whatever that will be).

Would also make sense to reword this slightly away from "These Workbox configuration options". It's more to highlight that in this context it won't be used, so maybe:

> The configuration options '${invalidConfig.join(',')}' will be ignored by `workbox-build.${methodName}`.

File: packages/workbox-build/src/lib/utils/warn-about-config.js (hunk 1 above)
Comment (jeffposnick) [reply packages/workbox-build/src/lib/utils/warn-about-config.js]: We're still aiming for Node 4 compatibility with these modules, so no `Array.prototype.includes()`: http://node.green/#ES2016-features-Array-prototype-includes

File: packages/workbox-build/src/lib/utils/warn-about-config.js (hunk 2 above)
Comment (jeffposnick) [reply packages/workbox-build/src/lib/utils/warn-about-config.js]: Done.

File: packages/workbox-build/src/lib/utils/warn-about-config.js (hunk 3 above)
Comment (jeffposnick) [reply packages/workbox-build/src/lib/utils/warn-about-config.js]: Folks might end up calling the methods indirectly via the Webpack or CLI, so I don't want to put `workbox-build.` in the message. But I'll reword along those lines.

**Pull request merged**